### PR TITLE
feat: resumable PPs download

### PR DESCRIPTION
**Pull request by hackaugusto**

### CODE REVIEW

File: lgn-worker/src/one-shot.rs [hunk 1]
```diff
@@ -73,10 +73,7 @@ async fn main() -> Result<()> {
     config.validate();
     let checksums = fetch_checksums(config.public_params.checksum_file_url()).await?;
 
-    let provers_manager = tokio::task::block_in_place(move || -> Result<ProversManager> {
-        ProversManager::new(&config, &checksums).context("while registering provers")
-    })
-    .context("creating prover managers")?;
```
Comment (hackaugusto) [inline lgn-worker/src/one-shot.rs:79]: This was needed because of the `reqwest::blocking::Client`. After this PR it is no longer used.

File: lgn-worker/src/manager.rs [hunk 2]
```diff
@@ -65,6 +72,7 @@ impl ProversManager {
                 &config.public_params.groth16_assets.r1cs_file,
                 &config.public_params.groth16_assets.pk_file,
             )
+            .await
             .context("initializing Large prover")?;
```
Comment (hackaugusto) [inline lgn-worker/src/manager.rs:76]: Making the download parallel should be trivial now, just need to add a `tokio::join!` here. I have not done that because of the memory use.

File: lgn-provers/src/params/mod.rs [hunk 3]
```diff
@@ -1,160 +1,247 @@
...
+            filepath.display(),
+        );
+
+        let mut stream = response.bytes_stream();
+        while let Some(data) = stream.next().await {
+            let data = data?;
+            file.write_all(&data).await?;
+            hasher.update_rayon(&data);
+            buf.extend(data);
+        }
```
Comment (hackaugusto) [inline lgn-provers/src/params/mod.rs:221]: A few notes:

- This does two copies:
   - `kernel -> use_space` when the data arrives
   - `user_space -> kernel -> file` when appending to the file
- This can eventually be changed to `kernel -> file` and `kernel -> user_space`, maybe it is has better performance for these large files

This also keeps a copy of the data in-memory, in the `buf`, ideally that should go away or maybe be mmaped to save memory.

Comment (delehef) [reply lgn-provers/src/params/mod.rs:221]: Frankly I think that mmapping at download-time does not make sense while we need to load in-RAM for plonky2 in any case.

File: lgn-worker/src/manager.rs (hunk 2 above)
Comment (delehef) [reply lgn-worker/src/manager.rs:76]: > Making the download parallel should be trivial now

I think that in any case, S3 will be serving fast enough to saturate the line, so I don't think that would save wall-clock time.

**Review by delehef — CHANGES_REQUESTED** [review]

File: lgn-provers/src/params/mod.rs [hunk 4]
```diff
@@ -1,160 +1,243 @@
...
-            local_param_filename.display()
-        )
-    })?;
-    std::fs::create_dir_all(current_param_dir).with_context(|| {
+    let mut filepath = PathBuf::from(param_dir);
+    filepath.push(file_name);
+
+    let param_dir = filepath
+        .parent()
+        .with_context(|| format!("Param directory can not be empty. param_dir: {}", param_dir))?;
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply lgn-provers/src/params/mod.rs:64]: I don't think that is correct. `PathBuf` doesn't hit the filesystem to validate if the directories or files exist, otherwise we wouldn't be able to use this to create files that doesn't exist yet. The only thing this does is to pop the last entry in the path:


```rust
fn main() {
    use std::path::PathBuf;

    let path = PathBuf::from("does_not_exist/inner/file.txt");
    println!("{:?}", path.parent());

    let path = PathBuf::from("file.txt");
    println!("{:?}", path.parent());
}
```

Prints:

```
Some("does_not_exist/inner")
Some("")
```

[the remark this answers is not in the document]
Comment (hackaugusto) [reply lgn-provers/src/params/mod.rs:64]: Actually, after that maybe we should also test if the value is `Some("")` I assume it would return `None` in that case.

[the remark this answers is not in the document]
Comment (delehef) [reply lgn-provers/src/params/mod.rs:64]: Indeed, you're right.

**Review by delehef — APPROVED** [review]

File: lgn-worker/src/manager.rs (hunk 2 above)
Comment (delehef) [reply lgn-worker/src/manager.rs:76]: I agree, but I would rank all that quite low in priority terms, for re-starting a worker is (hopefully) not a very frequent operation.
